executor/aggfuncs/func_group_concat.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggfuncs
+
+import (
+	"bytes"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+type baseGroupConcat4String struct {
+	baseAggFunc
+
+	sep string
+}
+
+func (e *baseGroupConcat4String) AppendFinalResult2Chunk(sctx sessionctx.Context, pr PartialResult, chk *chunk.Chunk) error {
+	p := (*partialResult4ConcatString)(pr)
+	if p.buffer == nil {
+		chk.AppendNull(e.ordinal)
+		return nil
+	}
+	chk.AppendString(e.ordinal, p.buffer.String())
+	return nil
+}
+
+type basePartialResult4GroupConcat struct {
+	buffer *bytes.Buffer
+}
+
+type partialResult4ConcatString struct {
+	basePartialResult4GroupConcat
+}
+
+type groupConcat4String struct {
+	baseGroupConcat4String
+}
+
+func (e *groupConcat4String) AllocPartialResult() PartialResult {
+	return PartialResult(new(partialResult4ConcatString))
+}
+
+func (e *groupConcat4String) ResetPartialResult(pr PartialResult) {
+	p := (*partialResult4ConcatString)(pr)
+	p.buffer = nil
+}
+
+func (e *groupConcat4String) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) (err error) {
+	p := (*partialResult4ConcatString)(pr)
+	v, isNull, valsBuf := "", false, make([]string, len(e.args))
+	for _, row := range rowsInGroup {
+		for i, l := 0, len(e.args); i < l; i++ {
+			v, isNull, err = e.args[i].EvalString(sctx, row)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if isNull {
+				continue
+			}
```

Comment:
how about this:
```go
func (e *groupConcat4String) writeSeprator(...) {
	if p.buffer == nil {
		p.buffer = &bytes.Buffer{}
	} else {
		p.buffer.WriteString(e.sep)
	}
}
```

Then for every parameter:
```go
writeSeprator := true
for arg := range e.args {
	...
	if writeSeprator {
		e.writeSeprator()
		writeSeprator = false
	}
	...
}
```

This can help us save the one memory copy: copies the string to the `valsBuf[i]`